### ISSUE TITLE
Incorrect reponses when using the OAuth authentication adapter

### ISSUE THREAD

**Issue by nuxwin** [zfcampus/zf-mvc-auth#108]
Due to the changes from #101feec0ec963726f7735c9daa23ba2904504d3b Oauth authentication adapter behavior is inconsistent and wrong.

### Clients are never challenged with proper 401 response

This is due to changes from #101feec0ec963726f7735c9daa23ba2904504d3b in which a 401 response is assumed only if the error parameter from the OAuth response is not null. While this assertion is true for an invalid, a malformed or an expired token, it is not the case for "initial challenge" (When no credential were provided).

Indeed, when an authentication request is checked by the OAUth server and if something goes wrong, we have only few response possibilities:

#### No credential were provided at all by the client

In such case, a 401 response with the appropriate WWW-Authenticate header is returned. This doesn't involve any error parameters in the WWW-Authenticate field.

#### Invalid, malformed or expired token

Same as above, but here, some error related parameters are added to the WWW-Authenticate header.

#### Insufficient scope

In such case, a 403 response is returned with a WWW-Authenticate header. Some error related parameters are also added to the WWW-Authentication header.

### Problem responses are not ApiProblemResponse

Apigility should be consistent when returning a response denoting a problem, whatever the authentication adapter in use.
Currently, in the OAUth authentication adapter case, the problem responses are not casted to ApiProblem response, resulting to an empty body.

### Expected behavior

- On authentication failure, or when no credentials were provided at all, client must be challenged with proper 401 response which include appropriate WWW-Authenticate header
- Response denoting problems must be casted to **ApiProblemResponse**

**Comment by nuxwin** [comment]
@jackdpeterson

Your opinion on this?

**Comment by jackdpeterson** [comment]
I agree with the above proposal. The problems with the current Apigility approach are well described above and are counter-intuitive for anyone who is developing a client-side application that relies on an appropriate HTTP status code. 403 is for insufficient privileges WHEN an identity is asserted. 401 is when one needs to assert an identity but hasn't or the credentials asserted are invalid.

I like the idea of differentiating between public and private API 'services' so long as it can be done at a per-method level (e.g., differentiate between GET and POST). I also like the idea of the early exit should no authentication step be required for performance reasons. 

This is definitely a Backwards compatibility break; however, in my opinion, it's a forward-compatibility enhancement and will promote wider-adoption for those that get to the point of developing an API that must be interacted with using numerous clients AND leveraging oauth or other complex authentication/authorization schemes.

Per RFC-2616:

10.4.2 401 Unauthorized

The request requires user authentication. The response MUST include a WWW-Authenticate header field (section 14.47) containing a challenge applicable to the requested resource. The client MAY repeat the request with a suitable Authorization header field (section 14.8). If the request already included Authorization credentials, then the 401 response indicates that authorization has been refused for those credentials. If the 401 response contains the same challenge as the prior response, and the user agent has already attempted authentication at least once, then the user SHOULD be presented the entity that was given in the response, since that entity might include relevant diagnostic information. HTTP access authentication is explained in "HTTP Authentication: Basic and Digest Access Authentication" [43].

AND

10.4.4 403 Forbidden

The server understood the request, but is refusing to fulfill it. Authorization will not help and the request SHOULD NOT be repeated. If the request method was not HEAD and the server wishes to make public why the request has not been fulfilled, it SHOULD describe the reason for the refusal in the entity. If the server does not wish to make this information available to the client, the status code 404 (Not Found) can be used instead.

I think the article here, http://robertlathanh.com/2012/06/http-status-codes-401-unauthorized-and-403-forbidden-for-authentication-and-authorization-and-oauth/ covers this topic very well and is consistent with my view on how 401 vs 403 should be treated.

**Comment by nuxwin** [comment]
@jackdpeterson 

Thank for your feedback. Well because apigility is event-driven based, We can provide a module that will change the default behavior. This module could replace default listeners by implementing guards

Doing this would be a good proof of concept and I think that @weierophinney could test the new behavior easily.

I'll start thinking about that module.

**Comment by weierophinney** [comment]
This repository has been closed and moved to laminas-api-tools/api-tools-mvc-auth; a new issue has been opened at https://github.com/laminas-api-tools/api-tools-mvc-auth/issues/9.
